Added IPTC reader [WIP]

vanshikaarora opened the pull request:
**Added Icafe and IPTC reader**

Fixes #2328 FBMD in picture metadata = Probable copyright violation

### Changes made
- Added the jar file for Icafe in the libs under Project root
- Added one class(MetadataEntry) of Jar file manually into our package(For it wasn't supported)
- Created a class ReadFBMD
- This class takes the image Url and extracts the IPTC
- We can Iterate over this IPTC to get the MetadataEntry which contains two strings key and value (Not sure whether this was the data that we were searching for or we'll figure it out later)

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,57 @@
+    ReadFBMD(Url url){
+        this.url=url;
+    }
+    public void processMetadata() throws IOException {
```
I understand that this is currently just for testing. But in the final PR, make this method return something that can be used.

vanshikaarora replied:
I  am sorry for that. I'll modify accordingly :+1:

vanshikaarora replied:
I'll make it return a boolean value whether the Image has FBMD data or not. Will that be fine?

maskaravivek replied:
Yes, that would be great.
- Does the processing take time? 
- Is it memory intensive?

vanshikaarora replied:
I'll verify all these properties closely and then revert back to you :)

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,57 @@
+
+public class ReadFBMD {
+    Url url;
+    ReadFBMD(Url url){
```
Ok I'll do that :+1:

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,57 @@
+            }
+        }
+    }
+    private void printMetadata(MetadataEntry entry, String indent, String increment) {
```
I'll remove that actually it was for 
```
logger.info(indent + entry.getKey() (StringUtils.isNullOrEmpty(entry.getValue())? "" : ": " + entry.getValue())) 
```
just to increase indent every time but we don't need it

nicolas-raoul commented:
If your pull request is a work in progress (not working yet) then I suggest editing the title and inserting `[WPI]` at the beginning of title :-)

maskaravivek commented:
@vanshikaarora You can do a `.toString` on the data of `IPTC` to get a string value. 

```
Metadata metadata = metadataMap.get(MetadataType.IPTC);
metadata.data.toString()
```

For the image I checked, I got this result: 

```
g��KLdQP7BXagEas5FxHch-(��bFBMD01000a9a0d00006d47000012750000937900002f7f0000ca8700009acb0000cbdb000040e6000048f20000237f0100
```

It contains what we are looking for: `FBMD`. 

<img width="1437" alt="screen shot 2019-02-05 at 12 48 45 am" src="https://user-images.githubusercontent.com/3069373/52231439-f8e3af80-28df-11e9-9a8f-f9e0f9832d8f.png">

maskaravivek commented:
Lets just show a warning dialog. We show a dialog when any of the image quality checks fails. Lets use the same dialog for showing this warning as well. 

Just jotting down the steps here so that you don't put in a lot of effort to add the warning show up in UI.

1. Define an error code for FBMD in `ImageUtils`: 

```
public static final int FILE_FBMD = -5;
```

2. Make `ReadFBMD:processMetadata` return FILE_FBMD value for image containing FBMD. 

3. And then add your `if` condition in `ImageUtils:getErrorMessageForResult`. 

It should automatically start showing this warning at appropriate places for `FBMD`.

vanshikaarora commented:
@maskaravivek I'll make these changes :+1:

vanshikaarora commented:
@maskaravivek Could you share the url that image?

For my image that field was empty

![screenshot from 2019-02-05 01-27-00](https://user-images.githubusercontent.com/34261945/52233522-54646c00-28e5-11e9-84a8-76d502245fb7.png)

vanshikaarora commented:
>Also, can you resolve the merge conflicts on this PR?

I'll do that shortly :+1:

vanshikaarora commented:
>Also, can you resolve the merge conflicts on this PR?

vanshikaarora commented:
>Also, can you resolve the merge conflicts on this PR?

I have resolved the merge conflicts

@maskaravivek @nicolas-raoul I have tested this on a few images here is one particular concern:
The byte array that all of us have posted in our screenshots get's reduced by it's length whenever I convert it into a byte array. And FBMD comes in the very beginning of the array
Please see that if the same happens with your images as well if we are sure that the same happens with all of them then I'll remove the iteration of byte array by just checking if the first four form the FBMD code

Other than this the PR is ready to be reviewed.
Kindly review the PR :)

vanshikaarora commented on app/build.gradle:
```diff
@@ -27,7 +28,7 @@ dependencies {
     implementation 'com.jakewharton.rxbinding2:rxbinding-design:2.1.1'
     implementation 'com.facebook.fresco:fresco:1.10.0'
-    
+
```
I apologise for that. I'll take care of it now onwards

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ImageProcessingService.java:
```diff
@@ -56,77 +67,93 @@ public ImageProcessingService(FileUtilsWrapper fileUtilsWrapper,
                 });
+
+        return Single.zip(zipResult, checkFBMD, (zip, fbmd) -> {
+            return zip | fbmd;
```
I have added the following:
`Timber.d("zip:"+zip+ "fbmd:"+zip);`

maskaravivek replied:
Nice! Anyways, timber supports string formatting so you can use: 
```
Timber.d("zip: %d fbmd: %d",zip, fbmd);
```

vanshikaarora replied:
>Nice! Anyways, timber supports string formatting so you can use:

```
Timber.d("zip: %d fbmd: %d",zip, fbmd);
```
Thanks for letting me know more :)

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ImageProcessingService.java:
```diff
@@ -56,77 +67,93 @@ public ImageProcessingService(FileUtilsWrapper fileUtilsWrapper,
-        Title title = uploadItem.getTitle();
-        if (title.isEmpty()) {
-            return Single.just(EMPTY_TITLE);
+        public Single<Integer> checkFBMD (String filePath){
```
Sorry for that 
I have removed it  :+1:

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ImageProcessingService.java:
```diff
@@ -56,77 +67,93 @@ public ImageProcessingService(FileUtilsWrapper fileUtilsWrapper,
-            return Single.just(EMPTY_TITLE);
+        public Single<Integer> checkFBMD (String filePath){
+            try {
+                return ReadFBMD.processMetadata(filePath);
```
Done :+1:

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,33 @@
+    public static Single<Integer> processMetadata(String path) throws IOException {
+        Map<MetadataType, Metadata> metadataMap = Metadata.readMetadata(path);
+        Metadata metadata = metadataMap.get(MetadataType.IPTC);
+        byte[] b = new byte[0];
```
Done :+1:

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,33 @@
+        } catch (NullPointerException e) {
+            return Single.just(ImageUtils.IMAGE_OK);
+        }
+        for (int i = 0; i < b.length; i++) {
```
I'll do that

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,47 @@
+
+@Singleton
+public class ReadFBMD {
+    public static ReadFBMD readFBMDInstance;
```
No need to have this. As you have already annotated the class with `@Singleton` you don't need to maintain a instance manually. :)

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,47 @@
+        return Single.just(ImageUtils.IMAGE_OK);
+    }
+
+    public static ReadFBMD getInstance(){
```
As commented above, no need for this. :)

vanshikaarora replied:
I have made this changes in the PR. Kindly review :)

maskaravivek commented:
@vanshikaarora Unit tests are failing. Can you fix them. :)

```
:app:testBetaDebugUnitTest
```

vanshikaarora commented:
Sure I'll do that

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java:
```diff
@@ -507,6 +507,39 @@ public boolean addWikidataEditTag(String revisionId) throws IOException {
         }).flatMapObservable(Observable::fromIterable);
     }
 
+    @Override
```
Oops! Did this get added by mistake?

vanshikaarora replied:
yes sorry i was actually adding this code for #2392

vanshikaarora commented:
>Unit tests are failing. Can you fix them. :)

@nicolas-raoul Shall I modify the UnitTest class?

nicolas-raoul commented:
Did you try running the unit tests locally on your computer?
Are some of them failing?
If yes, please check whether they also fail on master, and if not, have a look at the code of each failed test. By then you will probably know what is the problem better than us :-)

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,40 @@
+        } catch (NullPointerException e) {
+            return Single.just(ImageUtils.IMAGE_OK);
+        }
+            if (dataInBytes[0] == 70 && dataInBytes[1] == 66 && dataInBytes[2] == 77 && dataInBytes[3] == 68) {
```
This would look cryptic to a developer who arrives on the project, so better explain that these numbers are the codes for F, B, M, D.

vanshikaarora replied:
I have added a documentation just above this code :+1:

nicolas-raoul replied:
is used is tested to contain -> is used to check whether it contains
it's -> its
:-)

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/utils/ImageUtils.java:
```diff
@@ -43,6 +43,7 @@
     public static final int IMAGE_WAIT = -2;
     public static final int EMPTY_TITLE = -3;
     public static final int FILE_NAME_EXISTS = -4;
+    public static final int FILE_FBMD = -5;
```
Javadoc is probably needed here. Sorry that the other values have no Javadoc, but it is not to late to start doing things the good way :-)

vanshikaarora replied:
I have added the Javadoc for the variable

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/MetadataEntry.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * MetadataEntry.java
+ */
+public class MetadataEntry {
```
Please add Javadoc explaining what this class does, in what context.

vanshikaarora replied:
@nicolas-raoul This class wasn't present in the open source library but this class created some issues while downloading it as jar and then adding it as library. So i had created the class manually and it added it to the project.

Currently I see that we are not using this class anywhere in the project.

- Should I remove this class?

### Or

- Should I keep this class with a documentation so that it would won't create difficulty if we might need this in the future for the project

nicolas-raoul replied:
I suggest not adding this class, if it is not needed.

vanshikaarora replied:
Ok then I'll remove it :+1:

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/ImageProcessingService.java:
```diff
@@ -56,77 +68,94 @@ public ImageProcessingService(FileUtilsWrapper fileUtilsWrapper,
-        Title title = uploadItem.getTitle();
-        if (title.isEmpty()) {
-            return Single.just(EMPTY_TITLE);
+        public Single<Integer> checkFBMD(String filePath){
```
Please add method Javadoc explaining the whole reasoning behind this issue (what FBMD is, and how images from Facebook are often copyright violations).

vanshikaarora replied:
Also added a documentation here :+1:

vanshikaarora commented:
@nicolas-raoul I have added the documentation wherever necessary 

>If yes, please check whether they also fail on master, and if not, have a look at the code of each failed test.
The test weren't failing for the master, So probably the error was because of the function `validateImage` that was called on `imageProcessingService` in the `ImageProcessingServiceTest.kt`. The object `imageProcessingService` returns null for `readFBMD`

I have modified the unit test kindly review and share feedback :)

maskaravivek commented on app/src/main/java/fr/free/nrw/commons/utils/ImageUtils.java:
```diff
@@ -43,7 +43,11 @@
     public static final int IMAGE_WAIT = -2;
     public static final int EMPTY_TITLE = -3;
     public static final int FILE_NAME_EXISTS = -4;
+    public static final int FILE_FBMD = -5;
```
I didn't notice this earlier but is the error code same for both FILE_FBMD and NO_CATEGORY_SELECTED? Please make FILE_FBMD = -6 then

vanshikaarora replied:
I am sorry, even I didn't noticed that :(

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/utils/ImageUtils.java:
```diff
@@ -43,7 +43,11 @@
     static final int NO_CATEGORY_SELECTED = -5;
+    /*
+    * The parameter FILE_FBMD is returned from the class ReadFBMD if the uploaded image contains FBMD data else returns IMAGE_OK
+    * */
```
Please move this comment to immediately before the variable declaration, and make it a real Javadoc with `**`. That way, it will appear in IDE tooltips etc.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/upload/ReadFBMD.java:
```diff
@@ -0,0 +1,33 @@
+import io.reactivex.Single;
+import timber.log.Timber;
+
+public class ReadFBMD {
```
Is there any special reason why this class is not named FBMDReader? Using a noun is the norm. By the way, to make it faster to understand just by reading its name, I suggest renaming the class "FacebookMetadataReader" or "FacebookMetadataChecker" or something similar. Thanks!

maskaravivek commented:
@vanshikaarora is it working for you? For my image, it is not detecting FBMD. PFA logs: 

```
2019-02-10 16:19:29.062 27408-27408/? D/ImageProcessingService: Current image quality is -2
2019-02-10 16:19:29.062 27408-27408/? D/ImageProcessingService: Checking the validity of image
2019-02-10 16:19:29.063 27408-27408/? D/ImageProcessingService: Checking for duplicate image /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
2019-02-10 16:19:29.064 27408-27408/? D/ImageProcessingService: Checking for image geolocation /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
2019-02-10 16:19:29.064 27408-27408/? D/ImageProcessingService: Checking for dark image /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
2019-02-10 16:19:29.066 27408-27408/? D/ImageProcessingService: Checking for image title vivek maskara in full swing
2019-02-10 16:19:29.839 27408-28198/? D/ImageProcessingService: Result for duplicate image false
2019-02-10 16:19:30.824 27408-28198/? D/ImageProcessingService: Result for valid title is false
2019-02-10 16:19:30.825 27408-28198/? D/ImageProcessingService: Result for duplicate: 0, geo: 0, dark: 0, title: 0
2019-02-10 16:19:30.826 27408-28198/? D/ImageProcessingService: zip:0fbmd:0
```

Also, remove the .log file from the PR. I guess one error log file got added to git!

vanshikaarora commented:
> @vanshikaarora is it working for you? For my image, it is not detecting FBMD. PFA logs:
> 
> ```
> 2019-02-10 16:19:29.062 27408-27408/? D/ImageProcessingService: Current image quality is -2
> 2019-02-10 16:19:29.062 27408-27408/? D/ImageProcessingService: Checking the validity of image
> 2019-02-10 16:19:29.063 27408-27408/? D/ImageProcessingService: Checking for duplicate image /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
> 2019-02-10 16:19:29.064 27408-27408/? D/ImageProcessingService: Checking for image geolocation /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
> 2019-02-10 16:19:29.064 27408-27408/? D/ImageProcessingService: Checking for dark image /data/user/0/fr.free.nrw.commons.beta/cache/CommonsContributions/a98626bd-8ba3-4c22-b8d5-0c09e435ca4a.jpg
> 2019-02-10 16:19:29.066 27408-27408/? D/ImageProcessingService: Checking for image title vivek maskara in full swing
> 2019-02-10 16:19:29.839 27408-28198/? D/ImageProcessingService: Result for duplicate image false
> 2019-02-10 16:19:30.824 27408-28198/? D/ImageProcessingService: Result for valid title is false
> 2019-02-10 16:19:30.825 27408-28198/? D/ImageProcessingService: Result for duplicate: 0, geo: 0, dark: 0, title: 0
> 2019-02-10 16:19:30.826 27408-28198/? D/ImageProcessingService: zip:0fbmd:0
> ```
> Also, remove the .log file from the PR. I guess one error log file got added to git!

@maskaravivek Yes the EXIF data worked for me. All the screenshots that I have posted above are taken from Facebook downloaded images

Has it happened just once? Or for all the Facebook downloaded images

maskaravivek commented:
Will check tonight. Sorry, I checked just for one image.

maskaravivek commented:
Yes, it seems the check works but no warning dialog appeared for FBMD image. Anyways as its a very small change, I am raising a PR to show the warning dialog.

maskaravivek commented:
Here's the PR https://github.com/vanshikaarora/apps-android-commons/pull/4

vanshikaarora commented:
@maskaravivek Sorry for the long delay :(

I have pushed the current working code. Please check does this works for you :)

nicolas-raoul commented:
@vanshikaarora Congratulations! I guess you can remove "[WIP]" from this PR's title :-)
I will test the feature now.

vanshikaarora commented:
>@vanshikaarora Congratulations! I guess you can remove "[WIP]" from this PR's title :-)
I will test the feature now.

Thanks @nicolas-raoul :)

vanshikaarora commented:
>Thanks, @vanshikaarora. It works for me so I am going ahead and merging the PR. :)

Thanks for approving the PR @maskaravivek :)

>I recently added [another library](https://github.com/drewnoakes/metadata-extractor) for extracting EXIF data. Both libraries support EXIF and IPTC. So we should use either of these libraries. I will open another issue for this.

I saw that library and that's pretty good, I am starting to work on #97 so I believe for implementing it we need to be sure which library shall we be using. Do you have any particular opinion for the choice of library @maskaravivek @nicolas-raoul

nicolas-raoul commented:
The smallest one that handles both EXIF and IPTC, I would say :-)

On Mon, Feb 25, 2019, 21:11 Vanshika Arora <notifications@github.com> wrote:

> Thanks, @vanshikaarora <https://github.com/vanshikaarora>. It works for
> me so I am going ahead and merging the PR. :)
>
> Thanks for approving the PR @maskaravivek
> <https://github.com/maskaravivek> :)
>
> I recently added another library
> <https://github.com/drewnoakes/metadata-extractor> for extracting EXIF
> data. Both libraries support EXIF and IPTC. So we should use either of
> these libraries. I will open another issue for this.
>
> I saw that library and that's pretty good, I am starting to work on #97
> <https://github.com/commons-app/apps-android-commons/issues/97> so I
> believe for implementing it we need to be sure which library shall we be
> using. Do you have any particular opinion for the choice of library
> @maskaravivek <https://github.com/maskaravivek> @nicolas-raoul
> <https://github.com/nicolas-raoul>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/commons-app/apps-android-commons/pull/2383#issuecomment-466989000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGFBs7gKeQu9LkUHDkqXx5z7LNs0kpIks5vQ9L6gaJpZM4ac0lz>
> .
>

maskaravivek commented:
From what I checked both libraries add very little to the size of the final APK. It would be great if could use https://github.com/drewnoakes/metadata-extractor for both EXIF and IPTC. 

Initially, I tried using icafe for EXIF but I was unable to convert the date string to a proper `Date` object.

vanshikaarora commented:
>It would be great if could use https://github.com/drewnoakes/metadata-extractor for both EXIF and IPTC.

@maskaravivek the current PR uses Icafe so shall i use this replace icafe with metadata-extractor in the same PR or create a separate issue for it?

maskaravivek commented:
I have already merged this PR to master. Please create a separate issue for it and you can take it up if you want. :)

vanshikaarora commented:
>From what I checked both libraries add very little to the size of the final APK.

@maskaravivek have you successfully received IPTC using metadata extractor
```
metadata = ImageMetadataReader.readMetadata(new File(path));
            for (Directory directory : metadata.getDirectories()) {
                for (Tag tag : directory.getTags()) {
                    System.out.format("[%s] - %s = %s",
                            directory.getName(), tag.getTagName(), tag.getDescription());
                    System.out.println();
                }
                if (directory.hasErrors()) {
                    for (String error : directory.getErrors()) {
                        System.err.format("ERROR: %s", error);
                    }
                }
            }
```

I got the following output:
```
/System.out: [JPEG] - Compression Type = Progressive, Huffman
    [JPEG] - Data Precision = 8 bits
I/System.out: [JPEG] - Image Height = 1123 pixels
    [JPEG] - Image Width = 540 pixels
    [JPEG] - Number of Components = 3
    [JPEG] - Component 1 = Y component: Quantization table 0, Sampling factors 2 horiz/2 vert
I/System.out: [JPEG] - Component 2 = Cb component: Quantization table 1, Sampling factors 1 horiz/1 vert
    [JPEG] - Component 3 = Cr component: Quantization table 1, Sampling factors 1 horiz/1 vert
I/System.out: [JFIF] - Version = 1.1
    [JFIF] - Resolution Units = none
I/System.out: [JFIF] - X Resolution = 1 dot
    [JFIF] - Y Resolution = 1 dot
I/System.out: [JFIF] - Thumbnail Width Pixels = 0
    [JFIF] - Thumbnail Height Pixels = 0
I/System.out: [Huffman] - Number of Tables = 2 Huffman tables
    [File Type] - Detected File Type Name = JPEG
    [File Type] - Detected File Type Long Name = Joint Photographic Experts Group
I/System.out: [File Type] - Detected MIME Type = image/jpeg
    [File Type] - Expected File Name Extension = jpg
    [File] - File Name = 00675657-3f07-4328-a9de-8f5d2ef34e7b.jpg
    [File] - File Size = 38384 bytes
I/System.out: [File] - File Modified Date = Tue Feb 26 14:27:35 +05:30 2019
```

Also searching  for IPTC i got a null response:

![screenshot from 2019-02-26 14-55-33](https://user-images.githubusercontent.com/34261945/53401917-c442af80-39d6-11e9-8a23-b81c479695c0.png)

@nicolas-raoul @maskaravivek Can you please look into this :)
